netket/utils/semigroup.py

```diff
@@ -187,3 +186,145 @@ def __repr__(self):
+                group from the result.
+
+        Returns:
+            group: the permutation group with duplicate elements removed.
+            return_inverse: Indices to reconstruct the original group from the result.
+                Only returned if `return_inverse` is True.
+        """
+        result = np.unique(
+            self.to_array(),
+            axis=0,
+            return_index=True,
+            return_inverse=return_inverse,
+        )
+        group = PermutationGroup(
+            [self.elems[i] for i in sorted(result[1])], self.degree
+        )
+        if return_inverse:
+            return group, result[2]
+        else:
+            return group
+
+    def __inverse(self):
+        perm_array = self.to_array()
+        n_symm = len(perm_array)
+        inverse = np.zeros([n_symm], dtype=int)
+        for i, perm1 in enumerate(perm_array):
+            for j, perm2 in enumerate(perm_array):
+                perm_sq = perm1[perm2]
+                if np.all(perm_sq == np.arange(len(perm_sq))):
+                    inverse[i] = j
+
+        return inverse
+
+    def __product_table(self):
+        perms = self.to_array()
+        inverse = perms[self.inverse()].squeeze()
+        n_symm = len(perms)
+        product_table = np.zeros([n_symm, n_symm], dtype=int)
+
+        inv_t = inverse.transpose()
+        perms_t = perms.transpose()
+        inv_elements = perms_t[inv_t].reshape(-1, n_symm * n_symm).transpose()
+
+        perms = [HashableArray(element) for element in perms]
+        inv_perms = [HashableArray(element) for element in inv_elements]
+
+        inverse_index_mapping = {element: index for index, element in enumerate(perms)}
+
+        inds = [
+            (index, inverse_index_mapping[element])
+            for index, element in enumerate(inv_perms)
+            if element in inverse_index_mapping
+        ]
+
+        inds = np.asarray(inds)
+
+        product_table[inds[:, 0] // n_symm, inds[:, 0] % n_symm] = inds[:, 1]
+
+        return product_table
+
+    def inverse(self):
+        """
+        Returns indices of the involution of the PermutationGroup where the each element
+        is the inverse of the original element at the same position. If :code:`g = self[element]`
+        and :code:`h = self[self.inverse()][element]`, then :code:`gh = product(g, h)` will act
+        as the identity on the sites of the graph, i.e., :code:`np.all(gh(sites) == sites)`.
+        """
+        if self._inverse is None:
+            object.__setattr__(self, "_inverse", self.__inverse())
+
+        return self._inverse
+
+    def product_table(self):
+        """
+        Returns a product table over the group where the columns use the involution
+        of the group. If :code:`g = self[self.inverse()[element]]', :code:`h = self[element2]`
+        and code:`u = self[product_table()[element,element2]], we are
+        solving the equation u = gh
+        """
```

Comment:
* It's actually the rows that are reordered
* "Involution" doesn't really mean what we take it to mean, and in any case, plain language is better

```suggestion
        """
        Returns a table of g⁻¹h over the group. That is, if :code:`g = self[self.inverse()[element]]',
        :code:`h = self[element2]` and code:`u = self[product_table()[element,element2]], u = gh.
        """
```

Reply:
Actually, this might be even clearer
```suggestion
        """
        Returns a table of g⁻¹h over the group. That is, if :code:`g = self[element]',
        :code:`h = self[element2]`, and code:`u = self[product_table()[element,element2]], 
        then u = g⁻¹h.
        """
```